pkg/engine/lifecycletest/pulumi_test.go

```diff
@@ -5393,6 +5393,110 @@ func TestResourceNames(t *testing.T) {
 func TestSourcePositions(t *testing.T) {
 	t.Parallel()
 
+	loaders := []*deploytest.ProviderLoader{
+		deploytest.NewProviderLoader("pkgA", semver.MustParse("1.0.0"), func() (plugin.Provider, error) {
+			return &deploytest.Provider{
+				CreateF: func(urn resource.URN, inputs resource.PropertyMap, timeout float64,
+					preview bool,
+				) (resource.ID, resource.PropertyMap, resource.Status, error) {
+					return "created-id", inputs, resource.StatusOK, nil
+				},
+				ReadF: func(urn resource.URN, id resource.ID,
+					inputs, state resource.PropertyMap,
+				) (plugin.ReadResult, resource.Status, error) {
+					return plugin.ReadResult{Inputs: inputs, Outputs: state}, resource.StatusOK, nil
+				},
+				ConstructF: func(monitor *deploytest.ResourceMonitor,
+					typ, name string, parent resource.URN, inputs resource.PropertyMap,
+					options plugin.ConstructOptions,
+				) (plugin.ConstructResult, error) {
+					// The position of the root component should be replaced with the position of the corresponding call
+					// to RegisterResource in the user program (see compPos below).
+					urn, _, _, err := monitor.RegisterResource(tokens.Type(typ), name, false, deploytest.ResourceOptions{
+						Parent:         parent,
+						SourcePosition: "/component/source/positions#7,8",
+					})
+					require.NoError(t, err)
+
+					// The position of this resource should be elided, as it is not part of the user program.
+					_, _, _, err = monitor.RegisterResource("pkgA:m:typA", "resD", true, deploytest.ResourceOptions{
+						Parent:         urn,
+						SourcePosition: "/component/source/positions#9,10",
+					})
+					require.NoError(t, err)
+
+					err = monitor.RegisterResourceOutputs(urn, resource.PropertyMap{})
+					assert.NoError(t, err)
+
+					return plugin.ConstructResult{
+						URN:     urn,
+						Outputs: resource.PropertyMap{},
+					}, nil
+				},
+			}, nil
+		}),
+	}
+
+	const regPos = "/test/source/positions#1,2"
+	const readPos = "/test/source/positions#3,4"
+	const compPos = "/test/source/positions#5,6"
+	inputs := resource.PropertyMap{}
+	program := deploytest.NewLanguageRuntime(func(_ plugin.RunInfo, monitor *deploytest.ResourceMonitor) error {
+		_, _, _, err := monitor.RegisterResource("pkgA:m:typA", "resA", true, deploytest.ResourceOptions{
+			Inputs:         inputs,
+			SourcePosition: "file://" + regPos,
+		})
+		require.NoError(t, err)
+
+		_, _, err = monitor.ReadResource("pkgA:m:typA", "resB", "id", "", inputs, "", "", "file://"+readPos)
+		require.NoError(t, err)
+
+		_, _, _, err = monitor.RegisterResource("pkgA:m:typB", "resC", true, deploytest.ResourceOptions{
```

Comment:
Shouldn't custom be false here? It's a remote component, not a remote custom. I think this _happens_ to work today but could start triggering an error in the future if we get the engine to reject "custom=true,remote=true" registrations.